tests/frontend/test_parameters.py

```diff
@@ -206,3 +225,127 @@ def test_parameter_unary_negation(self, p):
         pp.val = p
         assert par_evaluate(-p) == pytest.approx(-p)
         assert par_evaluate(-pp) == pytest.approx(-p)
+
+
+def mock_run_prog(self, prog, **kwargs):
+    """Mock function that uses a global variable for constructing
+    the engine queue when running programs."""
+    global applied
+    for cmd in prog.circuit:
+        try:
+            cmd.op.apply(cmd.reg, self.backend, **kwargs)
+            applied.append(cmd)
+        except NotImplementedError:
+            for c in cmd.op._decompose(cmd.reg, **kwargs):
+                c.op.apply(c.reg, self.backend, **kwargs)
+                applied.append(c)
+    return applied
+
+
+class TestParameterTFIntegration:
+    """Test integration of the parameter handling system with
+    various gates and TensorFlow."""
+    pytest.importorskip("tensorflow", minversion="2.0")
+
+    @pytest.mark.parametrize("gate", [sf.ops.Dgate, sf.ops.Sgate, sf.ops.Coherent])
+    def test_single_mode_gate_complex_phase(self, backend, gate, monkeypatch):
+        """Test single mode gates with complex phase arguments"""
+        import tensorflow as tf
+
+        prog = sf.Program(1)
+
+        # create symbolic parameters
+        r = prog.params('r')
+        phi = prog.params('phi')
+
+        with prog as q:
+            gate(r, phi) | q[0]
+
+        # create numeric values
+        mapping = {'r': tf.Variable(0.1), 'phi': tf.Variable(0.1)}
+        prog.bind_params(mapping)
+
+        # verify bound parameters are correct
+        assert prog.free_params['r'].val is mapping['r']
+        assert prog.free_params['phi'].val is mapping['phi']
+
+        # assert executed program is constructed correctly
```

Comment:
From this on the test case could be split up into two: one for checking the parameters and one for checking that the executed program is constructed correctly.

Reply:
I've streamlined the tests a bit --- let me know if it looks better!

I didn't separate out the program and engine run into separate tests, however. This is partly because the program mapping is already being tested earlier in the file; this bug/intergration test only affects the engine run. The previous assertion on `prog.free_params` is just a sanity check.

Reply:
Looks better, thanks! :100: